datafusion/functions-window/src/ntile.rs

```diff
@@ -0,0 +1,192 @@
+// specific language governing permissions and limitations
+// under the License.
+
+use std::any::Any;
+use std::fmt::Debug;
+use std::sync::{Arc, OnceLock};
+
+use crate::utils::{get_scalar_value_from_args, get_unsigned_integer};
+use datafusion_common::arrow::array::{ArrayRef, UInt64Array};
+use datafusion_common::arrow::datatypes::{DataType, Field};
+use datafusion_common::{exec_err, DataFusionError, Result, ScalarValue};
+use datafusion_expr::window_doc_sections::DOC_SECTION_RANKING;
+use datafusion_expr::{
+    Documentation, Literal, PartitionEvaluator, Signature, Volatility, WindowUDFImpl,
+};
+use datafusion_functions_window_common::expr::ExpressionArgs;
+use datafusion_functions_window_common::field;
+use datafusion_functions_window_common::partition::PartitionEvaluatorArgs;
+use datafusion_physical_expr_common::physical_expr::PhysicalExpr;
+use field::WindowUDFFieldArgs;
+
+get_or_init_udwf!(
+    Ntile,
+    ntile,
+    "integer ranging from 1 to the argument value, dividing the partition as equally as possible"
+);
+
+pub fn ntile(arg: i64) -> datafusion_expr::Expr {
+    ntile_udwf().call(vec![arg.lit()])
+}
+
+#[derive(Debug)]
+pub struct Ntile {
+    signature: Signature,
+}
+
+impl Ntile {
+    /// Create a new `ntile` function
+    pub fn new() -> Self {
+        Self {
+            signature: Signature::uniform(
+                1,
+                vec![
+                    DataType::UInt64,
+                    DataType::UInt32,
+                    DataType::UInt16,
+                    DataType::UInt8,
+                    DataType::Int64,
+                    DataType::Int32,
+                    DataType::Int16,
+                    DataType::Int8,
+                ],
+                Volatility::Immutable,
+            ),
+        }
+    }
+}
+
+impl Default for Ntile {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+static DOCUMENTATION: OnceLock<Documentation> = OnceLock::new();
+
+fn get_ntile_doc() -> &'static Documentation {
+    DOCUMENTATION.get_or_init(|| {
+        Documentation::builder()
+            .with_doc_section(DOC_SECTION_RANKING)
+            .with_description(
+                "Integer ranging from 1 to the argument value, dividing the partition as equally as possible",
+            )
+            .with_syntax_example("ntile(expression)")
+            .with_argument("expression","An integer describing the number groups the partition should be split into")
+            .build()
+            .unwrap()
+    })
+}
+
+impl WindowUDFImpl for Ntile {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn name(&self) -> &str {
+        "ntile"
+    }
+
+    fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    fn expressions(&self, expr_args: ExpressionArgs) -> Vec<Arc<dyn PhysicalExpr>> {
+        parse_expr(expr_args.input_exprs(), expr_args.input_types())
+            .into_iter()
+            .collect::<Vec<_>>()
+    }
```

Comment:
Is this necessary?

Reply:
I think you can remove this completely and rely on the default implementation for `ntile`. From what I see `ntile` doesn't depend on the input expression in partition evaluator.

This implementation is required in `lead/lag` because it is possible that it can be called with `NULL`, and we try to deduce the type for that null if a default value argument exists. For `lead/lag` the types of the input expression and the default value needs to be the same in partition evaluator.